### PR TITLE
registry: Add more exhaustive search for fuzzy matching

### PR DESCRIPTION
**Pull request by jkozera**
Fixes 2 search cases not matching with fuzzy search:

1. "**ad**darg" not matching  "**a**rgparse.ArgumentParser.a**d**d_argument" in the Python docset,
2. "**te**strle" not matching "AWS.Elas**t**icTranscod**e**r.testRole" in the AWS JS docset.

Both were due to the 2nd char of needle being more than 8 chars from 1st match of 1st needle char in haystack (highlighted with **bold** above). Adding a simple recursive call to retry matching from 2nd match of 1st char fixes both cases.

It is slightly slower now - my simple performance test case of [all 2M symbols from all docsets](https://github.com/jkozera/zeal/blob/c30da109e8624b961e7b94f3f59f783b188c0d91/tests/py_unit/testscoring_build.py#L35) indicates about 20% slow down, from 1.2 seconds to 1.45 seconds on my box, on the "aaaaaaaaaa" query (which should be rather heavy, since it should involve scanning all the symbols multiple times due to lack of matches).

This investigation was inspired by the [recursive matching from lib_fts](https://github.com/forrestthewoods/lib_fts/blob/d541eb49b289a03320fb62150b8fd01364e22403/code/fts_fuzzy_match.h#L121), but for now I've decided to go with a simplified implementation, since in general it would be even slower. Preliminary testing with recursive call on each character match shows the slowdown would be like 40% instead of 20%, so maybe still feasible, but I'd like to try the faster implementation first. Especially since I'm also not sure if recursing on each match would actually improve the results - so if someone can find a reasonable test case which still doesn't work here, I'd reconsider. :)

### CODE REVIEW

File: src/libs/registry/docset.cpp [hunk 1]
```diff
@@ -639,6 +660,9 @@ static void matchFuzzy(const char *needle, int needleLength,
     *start = -1;
 
     int groupCount = 0;
+    int bestRecScore = -1;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Please do not use ambiguous shortenings. Recursive? Record?

File: src/libs/registry/docset.cpp [hunk 2]
```diff
@@ -647,9 +671,24 @@ static void matchFuzzy(const char *needle, int needleLength,
 
         while (j < haystackLength) {
             if (needle[i] == haystack[j++]) {
-                if (*start == -1)
+                if (*start == -1) {
                     *start = j;  // first matched char
 
+                    // try starting the search later in case the first character occurs again later
+                    int recStart, recLength, recScore;
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Same as above.

File: src/libs/registry/docset.cpp [hunk 3]
```diff
@@ -647,9 +671,24 @@ static void matchFuzzy(const char *needle, int needleLength,
...
+                    matchFuzzy(
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Don't break the line immediately after parenthesis, there's quite a lot of space on the same line (up to 100 characters).

File: src/libs/registry/docset.cpp [hunk 4]
```diff
@@ -674,10 +713,23 @@ static void matchFuzzy(const char *needle, int needleLength,
...
+                *start = bestRecStart;
+                *length = bestRecLength;
+            } else {
+                *start = -1;
+            }
             return;
         }
     }
+
+    int score = scoreFuzzy(haystack, *start, *length);
```
Comment (trollixx) [inline src/libs/registry/docset.cpp]: Why re-score again? This case should've been considered by now. Or not?

Comment (jkozera) [inline src/libs/registry/docset.cpp]: Above only the cases of some `recStart > *start` were considered - here we compare all the later-starting scores against the 'earliest' score. For example, when searching for in "foobaz.foobar" could yield two different scores - matching "[**foo**]baz.foo[**bar**]" or "foobaz.[**foobar**]" - in the latter case of exact match the score would be larger.

I hope to address the other comments later today.

**Pull request merged**